### PR TITLE
Master l10n ch isr oco

### PR DESCRIPTION
**Pull request by oco-odoo**
Swiss localization : postal references and ISR support (without ISR report layout).

### CODE REVIEW

File: addons/l10n_ch/models/res_bank.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    @api.depends("zip", "city")
+    def _compute_location_line(self):
+        for record in self:
+            if record.zip and record.city:
+                record.location_line = record.zip + ' ' + record.city
+
+class ResPartnerBank(models.Model):
+    _inherit = 'res.partner.bank'
+
+    postal_reference = fields.Char(help='The ISR number of the company within the bank', compute='_compute_postal_reference')
```
Comment (antonylesuisse) [inline addons/l10n_ch/models/res_bank.py]: l10n_ch_postal

File: addons/l10n_ch/models/res_company.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+class Company(models.Model):
+    _inherit = "res.company"
+
+    isr_print_bank_location = fields.Boolean(default=False, help='Boolean option field indicating whether or not the alternate layout (the one printing bank name and address) must be used when generating an ISR.')
```
Comment (antonylesuisse) [inline addons/l10n_ch/models/res_company.py]: l10n_ch_isr....

File: addons/report/static/src/js/qwebactionmanager.js [hunk 3]
```diff
@@ -98,12 +98,27 @@ ActionManager.include({
...
+                    //TODO DOC !!
+                    var treated_actions = [];
+                    var current_action = action;
+                    do {
+                        report_urls = make_report_url(current_action);
+                        response = [
+                            report_urls['qweb-pdf'],
+                            action.report_type, //The 'root' report is considered the maine one, so we use its type for all the others.
+                        ];
+                        trigger_download(self.session, response, c, current_action, options);
```
Comment (antonylesuisse) [inline addons/report/static/src/js/qwebactionmanager.js]: Check with jerry

File: addons/l10n_ch/models/account_invoice.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                invoice_issuer_ref = re.sub('^0*','',record.partner_bank_id.l10n_ch_postal)
+                invoice_issuer_ref = invoice_issuer_ref.ljust(l10n_ch_ISR_NUMBER_ISSUER_LENGTH, '0')
+                invoice_ref = re.sub('[^\d]', '', record.number)
+                invoice_ref = invoice_ref[-l10n_ch_ISR_NUMBER_ISSUER_LENGTH:] #We only keep the last digits of the sequence number if it is too long
+                internal_ref = zfill(invoice_ref, l10n_ch_ISR_NUMBER_LENGTH - l10n_ch_ISR_NUMBER_ISSUER_LENGTH)
+                record.l10n_ch_isr_number = invoice_issuer_ref + internal_ref
+
+    @api.depends('number', 'partner_bank_id.l10n_ch_postal')
+    def _compute_l10n_ch_isr_number_spaced(self):
+        for record in self:
```
Comment (qdp-odoo) [inline addons/l10n_ch/models/account_invoice.py]: merge with _compute_l10n_ch_isr_number()

File: addons/l10n_ch/models/account_invoice.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    def _compute_l10n_ch_isr_valid(self):
+        for record in self:
+            record.l10n_ch_isr_valid = record.type == 'out_invoice' and\
+                record.number and \
+                record.l10n_ch_isr_postal and \
+                record.partner_bank_id and \
+                record.partner_bank_id.l10n_ch_postal and \
+                record.currency_id.name in ['EUR', 'CHF']
+
+    def split_total_amount(self):
```
Comment (qdp-odoo) [inline addons/l10n_ch/models/account_invoice.py]: remove method, it is not used anymore

File: addons/l10n_ch/models/mail_template.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        """
+        rslt = super(MailTemplate, self).generate_email(res_ids, fields)
+
+        multi_mode = True
+        if isinstance(res_ids, (int, long)):
+            res_ids = [res_ids]
+            multi_mode = False
+
+        res_ids_to_templates = self.get_email_template(res_ids)
+        for res_id in res_ids:
```
Comment (qdp-odoo) [inline addons/l10n_ch/models/mail_template.py]: in the same fashion

File: odoo/tools/float_utils.py [hunk 7]
```diff
@@ -150,6 +151,18 @@ def float_repr(value, precision_digits):
...
 
+def float_split(value, precision_digits):
+    """ Splits the float given as 'value' argument, returning a tupe containing
+    (as integers) its number of units (as element 0), and the digits after the dot
+    in its representation (as element 1), limited to the provided number of
+    precision digits passed in arguments.
+    """
+    value = float_round(value, precision_digits=precision_digits)
+    units = math.floor(value)
+    cents = math.floor(float_round((value - units) * 10**precision_digits, precision_digits=2))
```
Comment (qdp-odoo) [inline odoo/tools/float_utils.py]: I don't think we need the extra float_round, here.. do we?

File: addons/l10n_ch/models/res_company.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+class Company(models.Model):
+    _inherit = "res.company"
+
+    l10n_ch_isr_print_bank_location = fields.Boolean(default=False, help='Boolean option field indicating whether or not the alternate layout (the one printing bank name and address) must be used when generating an ISR.')
```
Comment (qdp-odoo) [inline addons/l10n_ch/models/res_company.py]: string=??

File: addons/l10n_ch/tests/test_l10n_ch_isr.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+            return self.env['res.partner.bank'].create({
+                'acc_number': number
+            })
+
+    def test_isr_not_generated(self, invoice=None): #invoice parameter is mandatory ; this declaration is used for compliance with the tests
+        """ Prints the given invoice and tests that no ISR generation is triggered. """
+        if not invoice:
+            return
+
+        action_to_execute = invoice.invoice_print()
```
Comment (qdp-odoo) [inline addons/l10n_ch/tests/test_l10n_ch_isr.py]: next_action = invoice.invoice_print()

File: addons/l10n_ch/tests/test_l10n_ch_isr.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        return invoice
+
+    def create_test_account(self, number=None):
+        """ Generates a test res.partner.bank. """
+        if number:
+            return self.env['res.partner.bank'].create({
+                'acc_number': number
+            })
+
+    def test_isr_not_generated(self, invoice=None): #invoice parameter is mandatory ; this declaration is used for compliance with the tests
```
Comment (qdp-odoo) [inline addons/l10n_ch/tests/test_l10n_ch_isr.py]: rename this method and remove default value

**Comment by qdp-odoo** [comment]
l10n_ch ISR and postal bank accounts support has landed on saas-15 already at revision https://github.com/odoo/odoo/commit/9d42807ce3ced870aab6c8f70d2107e83cdcba11 (and previous revision).

This PR should have been closed at that time but was forgotten

**Pull request closed without merging**